batch/batch/driver/instance_collection/job_private.py

```diff
@@ -322,15 +323,26 @@ async def user_runnable_jobs(user, remaining):
             ):
                 async for record in self.db.select_and_fetchall(
                     '''
-SELECT jobs.job_id, jobs.spec, jobs.cores_mcpu, COALESCE(SUM(instances.state IS NOT NULL AND
-  (instances.state = 'pending' OR instances.state = 'active')), 0) as live_attempts
-FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_inst_coll_cancelled)
-LEFT JOIN attempts ON jobs.batch_id = attempts.batch_id AND jobs.job_id = attempts.job_id
-LEFT JOIN instances ON attempts.instance_name = instances.name
-WHERE jobs.batch_id = %s AND jobs.state = 'Ready' AND always_run = 1 AND jobs.inst_coll = %s
-GROUP BY jobs.job_id, jobs.spec, jobs.cores_mcpu
-HAVING live_attempts = 0
-LIMIT %s;
+WITH ready_jobs AS (
+  SELECT jobs.batch_id, jobs.job_id, jobs.spec, jobs.cores_mcpu, COALESCE(SUM(instances.state IS NOT NULL AND
+    (instances.state = 'pending' OR instances.state = 'active')), 0) as live_attempts
+  FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_inst_coll_cancelled)
+  LEFT JOIN attempts ON jobs.batch_id = attempts.batch_id AND jobs.job_id = attempts.job_id
+  LEFT JOIN instances ON attempts.instance_name = instances.name
+  WHERE jobs.batch_id = %s AND jobs.state = 'Ready' AND always_run = 1 AND jobs.inst_coll = %s
+  GROUP BY jobs.job_id, jobs.spec, jobs.cores_mcpu
+  HAVING live_attempts = 0
+  LIMIT %s
+)
+SELECT ready_jobs.*, regions
+FROM ready_jobs
+LEFT JOIN (
+  SELECT ready_jobs.batch_id, ready_jobs.job_id, JSON_ARRAYAGG(region) AS regions
+  FROM ready_jobs
+  INNER JOIN job_regions ON ready_jobs.batch_id = job_regions.batch_id AND ready_jobs.job_id = job_regions.job_id
+  LEFT JOIN region_ids ON job_regions.region_id = region_ids.region_id
+  GROUP BY ready_jobs.batch_id, ready_jobs.job_id
+) AS regions ON ready_jobs.batch_id = regions.batch_id AND ready_jobs.job_id = regions.job_id;
```

Comment:
The issue is that with a `LEFT JOIN` and `JSON_ARRAYAGG`, the result set returned is `[null]`. I thought that was confusing and wasn't sure whether the performance hit from how I wrote the subquery would matter. If it does matter, then I can change how we handle the result set here and check for `[None]` instead of `None`.

Reply:
I'm not sure I follow what you mean by `[null]`. Wouldn't the case be that if a job has no specified regions you'd get records like:
```
[(123, 1, ... spec ..., 1000, ... live_attempts ..., None),
 ...]
```

Reply:
wow, wtf...

Reply:
Ugh right left join

Reply:
So a reasonable fix here is `if(region is NULL, json_array(), json_arrayagg(region))`.

I concede I don't know the performance impact but it seems generally a good idea to not unnecessarily introduce subqueries.